File: generator/plugins/express/templates/src/migrate.js
'use strict';

// A tiny, deterministic migration runner (the Express equivalent of Flyway).
// It applies every migrations/V*.sql file in filename order exactly once,
// recording applied versions in a schema_migrations table. Standard SQL; no
// platform-specific markers (Laws 19-21).
const fs = require('fs');
const path = require('path');
const pool = require('./db');

const MIGRATIONS_DIR = path.join(__dirname, '..', 'migrations');

async function migrate() {
  await pool.query(
    'CREATE TABLE IF NOT EXISTS schema_migrations (version VARCHAR(255) PRIMARY KEY, applied_at __DB_SCHEMA_MIGRATIONS_TS__)',
  );
  const files = fs
    .readdirSync(MIGRATIONS_DIR)
    .filter((f) => f.endsWith('.sql'))
    .sort();
  for (const file of files) {
    const version = file;
    const { rowCount } = await pool.query('SELECT 1 FROM schema_migrations WHERE version = $1', [version]);
    if (rowCount > 0) continue;
    const sql = fs.readFileSync(path.join(MIGRATIONS_DIR, file), 'utf8');
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      await client.query(sql);
      await client.query('INSERT INTO schema_migrations (version) VALUES ($1)', [version]);
      await client.query('COMMIT');
      console.log('Applied migration ' + version);
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  }
}

module.exports = migrate;

File: generator/plugins/go/templates/internal/migrate/migrate.go
// THRAKSHA-owned shell. A tiny, deterministic migration runner (the Go equivalent
// of Flyway).
//
// It applies every migrations/V*.sql file in filename order exactly once, recording
// applied versions in a schema_migrations table. Standard SQL; no platform-specific
// markers (Laws 19-21). The bookkeeping timestamp type comes from the database
// provider, so it is correct for whichever database was selected.
package migrate

import (
	"database/sql"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const migrationsDir = "migrations"

// Run applies all pending migrations in order.
func Run(db *sql.DB) error {
	if _, err := db.Exec(
		"CREATE TABLE IF NOT EXISTS schema_migrations (version VARCHAR(255) PRIMARY KEY, applied_at __DB_SCHEMA_MIGRATIONS_TS__)",
	); err != nil {
		return err
	}

	entries, err := os.ReadDir(migrationsDir)
	if err != nil {
		return err
	}
	var files []string
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".sql") {
			files = append(files, e.Name())
		}
	}
	sort.Strings(files)

	for _, name := range files {
		var one int
		err := db.QueryRow("SELECT 1 FROM schema_migrations WHERE version = __DB_PH1__", name).Scan(&one)
		if err == nil {
			continue // already applied
		}
		if err != sql.ErrNoRows {
			return err
		}

		content, err := os.ReadFile(filepath.Join(migrationsDir, name))
		if err != nil {
			return err
		}
		tx, err := db.Begin()
		if err != nil {
			return err
		}
		if _, err := tx.Exec(string(content)); err != nil {
			_ = tx.Rollback()
			return err
		}
		if _, err := tx.Exec("INSERT INTO schema_migrations (version) VALUES (__DB_PH1__)", name); err != nil {
			_ = tx.Rollback()
			return err
		}
		if err := tx.Commit(); err != nil {
			return err
		}
		log.Printf("Applied migration %s", name)
	}
	return nil
}
